ops/lib/tcp_load_balancer/tcp_lb.py

```diff
@@ -0,0 +1,241 @@
+    to expose its data. If a load-balancer is highly-available, there will be multiple of those
+    events fired as instances of the load-balancer are observed by the member.
+    '''
+    pass
+
+
+class TcpLoadBalancerEvents(EventSetBase):
+
+    '''Events emitted by the TcpLoadBalancer class.'''
+    lb_available = EventSource(LBAvailable)
+
+
+class TcpLoadBalancer(Object):
+
+    on = TcpLoadBalancerEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm, relation_name, lb_algorithm=None):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._lb_algorithm = lb_algorithm
+
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event):
+        self.on.lb_available.emit()
+
+    @property
+    def _relation(self):
+        # TODO: there could be multiple independent reverse proxies in theory, address that later.
+        return self.model.get_relation(self._relation_name)
+
+    def expose_member(self, member, listener, health_monitor):
+        our_unit_data = self._relation.data[self.model.unit]
+        if member.address is None:
+            addr = self.model.get_binding(self._relation).network.ingress_address
+            member.set_address(addr)
+
+        our_unit_data['member'] = pickle.dumps(member, 0).decode('utf-8')
+        if self.model.unit.is_leader():
+            our_app_data = self._relation.data[self.model.app]
+            our_app_data['listener'] = pickle.dumps(listener, 0).decode('utf-8')
+            # A monitor for a pool of members.
+            our_app_data['health_monitor'] = pickle.dumps(health_monitor, 0).decode('utf-8')
+            our_app_data['lb_algorithm'] = self._lb_algorithm
+
+
+class Listener:
+
+    '''Listeners determine how load-balancer front-ends are configured.
+    '''
+
+    def __init__(self, name, port):
+        self.name = name
+        self.port = port
+
+
+class Member:
+
+    '''Members describe the details about a particular backend service instance.
+    '''
+
+    def __init__(self, name, port, *, address=None, monitor_port=None, weight=None,
+                 data_timeout=None):
+        self.name = name
+        self.port = port
+        self.address = address
+        self.monitor_port = monitor_port
+        self.weight = weight
+        self.data_timeout = data_timeout
+
+    def set_address(self, address):
+        self.address = address
+
+
+class HealthMonitor:
+
+    '''Health-monitors provide parameters for regular health-checking operations.
+    '''
+
+    def __init__(self, *, delay=None, timeout=None, max_retries=None, max_retries_down=None):
+        self.delay = delay
+        self.timeout = timeout
+        self.max_retries = max_retries
+        self.max_retries_down = max_retries_down
+
+
+class HTTPHealthMonitor(HealthMonitor):
```

Comment:
Earlier you use title case convention for initialisms, e.g.  `Tcp`. Recommend keeping these consistent.